support-frontend/assets/pages/contributions-landing/components/StripeCardForm/StripeCardForm.jsx

```diff
@@ -151,23 +192,66 @@ class CardForm extends Component<PropTypes, StateTypes> {
     });
   };
 
+  // Creates a new setupIntent upon recaptcha verification
+  setupRecaptchaCallback = () => {
+    window.grecaptcha.render('robot_checkbox', {
+      sitekey: window.guardian.v2recaptchaPublicKey,
+      callback: (token) => {
+        this.props.setStripeRecaptchaVerified(true);
+
+        fetch(
+          '/stripe/create-setup-intent/recaptcha',
+          requestOptions(
+            { token, stripePublicKey: this.props.stripeKey },
+            'same-origin',
+            'POST',
+            this.props.csrf,
+          ),
+        )
+          .then(response => response.json())
+          .then((json) => {
+            if (json.client_secret) {
+              this.props.setStripeSetupIntentClientSecret(json.client_secret);
+            } else {
+              throw new Error(`Missing client_secret field in server response: ${json}`);
+            }
+          })
+          .catch((err) => {
+            logException(`Error getting Setup Intent client_secret from /stripe/create-setup-intent/recaptcha: ${err}`);
+            this.props.paymentFailure('internal_error');
+            this.props.setPaymentWaiting(false);
+          });
+      },
+    });
+  };
+
   setupRecurringHandlers(): void {
     this.props.setCreateStripePaymentMethod(() => {
       this.props.setPaymentWaiting(true);
 
-      fetchJson(
-        window.guardian.stripeSetupIntentEndpoint,
-        requestOptions({ publicKey: this.props.stripeKey }, 'omit', 'POST', null),
-      ).then((result) => {
-        if (result.client_secret) {
-          this.handleCardSetupForRecurring(result.client_secret);
-        } else {
-          throw new Error(`Missing client_secret field in response from ${window.guardian.stripeSetupIntentEndpoint}`);
+      if (recaptchaEnabled(this.props.countryGroupId, this.props.contributionType)) {
+        /* Recaptcha verification is required for setupIntent creation.
+        If setupIntentClientSecret is ready then complete the payment now.
+        If setupIntentClientSecret is not ready then componentDidUpdate will complete the payment when it arrives. */
+        if (this.props.setupIntentClientSecret) {
```

Comment:
Why did the race condition come back into play?

Reply:
We request the client_secret when the user completes the captcha. Recaptcha immediately displays a green tick, but it may be some time before the client_secret is returned by our backend.

We then have 2 choices:
1. disable the Contribute button until the client_secret is ready (not great UX).
2. allow the user to click Contribute, bring up the usual spinner, and when the client_secret arrives complete the payment.

We choice the 2nd option because the UX is much better.